### PR TITLE
[test_fgnhg] Test enhancement for link operational change and matchmode changes to fine grained ECMP

### PR DESCRIPTION
**Pull request by anish-n**
### Description of PR

Summary:
Enhance test_fgnhg with link operational change testing. 

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

#### What is the motivation for this PR?
HLD update: https://github.com/Azure/SONiC/pull/693

#### How did you do it?
Change link operational state in the test and validate that ECMP adjusts to account for it

#### How did you verify/test it?
By running the test via pytest/ptf

#### Any platform specific information?
Applicable to all platforms which support fine grained ecmp

### Documentation 
https://github.com/Azure/SONiC/pull/693

### CODE REVIEW

File: tests/ecmp/test_fgnhg.py
Comment (kktheballer) [inline tests/ecmp/test_fgnhg.py:197]: Can we say "shutdown_link" instead of bank_0_port[0]?

File: tests/ecmp/test_fgnhg.py
Comment (kktheballer) [inline tests/ecmp/test_fgnhg.py:321]: nit: Spacing adjustment here maybe?

File: tests/ecmp/test_fgnhg.py
Comment (kktheballer) [inline tests/ecmp/test_fgnhg.py:350]: nit again: spacing

File: tests/ecmp/test_fgnhg.py
Comment (anish-n) [reply tests/ecmp/test_fgnhg.py:197]: good catch, will fix

File: tests/ecmp/test_fgnhg.py
Comment (anish-n) [reply tests/ecmp/test_fgnhg.py:321]: Which part of spacing looked off here?

File: tests/ecmp/test_fgnhg.py
Comment (anish-n) [reply tests/ecmp/test_fgnhg.py:350]: Which part of spacing looked off here?

**Review by kktheballer — APPROVED** [review]

**Comment by anish-n** [comment]
@nazariig please can you help review?

File: tests/ecmp/test_fgnhg.py
Comment (vivekreddynv) [inline tests/ecmp/test_fgnhg.py:647]: cleanup invocation in setup_teardown has to be changed to cleanup(duthost, ptfhost)

File: ansible/roles/test/files/ptftests/fg_ecmp_test.py
Comment (nazariig) [inline ansible/roles/test/files/ptftests/fg_ecmp_test.py:86]: @anish-n pytest.fail ?

Comment (anish-n) [reply ansible/roles/test/files/ptftests/fg_ecmp_test.py:86]: This is in the ptf test, I don't believe this is using pytest and has the pytest.fail available. Regardless, the exception will be reported upto the calling pytest, and it will show the failure there. Hope that makes sense.

File: tests/ecmp/test_fgnhg.py
Comment (anish-n) [reply tests/ecmp/test_fgnhg.py:647]: Can you please provide some more details as to what you are suggesting for modification here?

**Review by nazariig — APPROVED** [review]

File: tests/ecmp/test_fgnhg.py
Comment (abdosi) [inline tests/ecmp/test_fgnhg.py:470]: can we optimize `ptf_runner` this where we make it as fixture and pass only information needed for given case. Most of arguments seems fixed/constant in each case. 

Eg: everflow/test_everflow_testbed.py to check

File: tests/ecmp/test_fgnhg.py
Comment (abdosi) [inline tests/ecmp/test_fgnhg.py:423]: we can use API `get_vtysh_cmd_for_namespace()`

File: tests/ecmp/test_fgnhg.py
Comment (abdosi) [inline tests/ecmp/test_fgnhg.py:618]: why we need this local API ? `duthost.command()` should work

File: tests/ecmp/test_fgnhg.py
Comment (abdosi) [inline tests/ecmp/test_fgnhg.py:681]: can this be module level fixture with param ? We can do clean up as part of fixture. Not able to see where cleanup is happening.

**Review by abdosi — CHANGES_REQUESTED** [review]

File: tests/ecmp/test_fgnhg.py
Comment (anish-n) [reply tests/ecmp/test_fgnhg.py:618]: This is used for logging purposes, configure_dut calls duthost.command, except it logs prior to running durthost.command

File: tests/ecmp/test_fgnhg.py
Comment (anish-n) [reply tests/ecmp/test_fgnhg.py:681]: common_setup_teardown in line 678 does cleanup after test is completed. This is a similar model to what is done in a few other tests as well, eg: test_bgp_speaker.

File: tests/ecmp/test_fgnhg.py
Comment (anish-n) [reply tests/ecmp/test_fgnhg.py:423]: Can this improvement be taken as part of next iteration of the test? I would prefer having this PR checked in due to the length of time it has been open. The next PR will be opened shortly after this by Kavin.

File: tests/ecmp/test_fgnhg.py
Comment (anish-n) [reply tests/ecmp/test_fgnhg.py:470]: Can this improvement be taken as part of next iteration of the test? I would prefer having this PR checked in due to the length of time it has been open. The next PR will be opened shortly after this by Kavin.

Comment (abdosi) [reply tests/ecmp/test_fgnhg.py:470]: @anish-n can we merge next change also as part of this PR and optimize ? Wondering what is blocked on merge on this PR ?

Comment (anish-n) [reply tests/ecmp/test_fgnhg.py:470]: There is a different author for that PR, he already has an open PR for minigraph parsing enhancement which needs to go in after this https://github.com/Azure/sonic-mgmt/pull/2500, further, warm boot test case will also be added by Kavin and included in follow up PR.

Comment (abdosi) [reply tests/ecmp/test_fgnhg.py:470]: I think these are small enhancements. so better we can do now and then do not need follow up.

Comment (anish-n) [reply tests/ecmp/test_fgnhg.py]: Makes sense, have optimized PTF runner and now perform this part via a helper function, thanks for the comment!

**Review by wangxin — COMMENTED** [review]
Can you also fix the LGTM alerts?

File: tests/ecmp/test_fgnhg.py
Comment (anish-n) [reply tests/ecmp/test_fgnhg.py:349]: Added the same, thanks, vtysh_base_cmd is now formed via get_vtysh_cmd_for_namespace

File: tests/ecmp/test_fgnhg.py
Comment (anish-n) [reply tests/ecmp/test_fgnhg.py]: Thanks for the review, added that

File: tests/ecmp/test_fgnhg.py
Comment (anish-n) [reply tests/ecmp/test_fgnhg.py]: Good catch, thanks, fixed that

**Comment by anish-n** [comment]
> Can you also fix the LGTM alerts?

Fixed LGTM errors, the remaining one on ```result of integer division may be truncated``` is fine from the test logic perspective and may be ignored.

**Review by wangxin — APPROVED** [review]

**Review by abdosi — APPROVED** [review]

**Comment by abdosi** [comment]
can you fix LGTM ?

**Comment by anish-n** [comment]
> can you fix LGTM ?

Have fixed most LGTM errors, the remaining one on result of integer division may be truncated is fine from the test logic perspective and can be ignored?

**Comment by anish-n** [comment]
@abdosi , @wangxin can you please help merge the PR if it is good to go?

**Pull request merged**
